clash-cores/src/Clash/Cores/Xilinx/DcFifo/Explicit.hs

```diff
@@ -0,0 +1,291 @@
+{-|
+Copyright  :  (C) 2022, Google Inc,
+License    :  BSD2 (see the file LICENSE)
+Maintainer :  QBayLogic B.V. <devops@qbaylogic.com>
+
+Support for the [FIFO Generator v13.2](https://docs.xilinx.com/v/u/en-US/pg057-fifo-generator).
+
+
+-}
+
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE QuasiQuotes #-}
+
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+
+module Clash.Cores.Xilinx.DcFifo.Explicit
+  ( -- * Instantiating IP
+    dcFifo
+  , XilinxFifo(..)
+
+    -- * Customizing IP
+  , DcConfig(..)
+  , DcImplementation(..)
+  , ReadMode(..)
+  , defConfig
+
+    -- * Helper type aliases
+  , ResetBusy
+  , Full
+  , Empty
+  , DataCount
+  ) where
+
+import           Clash.Annotations.Primitive (Primitive (InlineYamlPrimitive))
+import           Clash.Explicit.Prelude
+import           Clash.Signal.Internal       (Signal (..))
+
+import           Data.Bifunctor              (second)
+import           Data.String.Interpolate     (i)
+
+-- We want constructor names to correspond 1:1 to Xilinx options it ease
+-- blackbox implementation.
+{-# HLINT ignore "Use camelCase" #-}
+
+type ResetBusy = Bit
+type Full = Bool
+type Empty = Bool
+type DataCount n = Unsigned n
+
+data ReadMode
+  = Standard_FIFO
+  | First_Word_Fall_Through
+  deriving (Show, Generic)
+
+data DcImplementation
+  = Independent_Clocks_Block_RAM
+  | Independent_Clocks_Distributed_RAM
+  | Independent_Clocks_Builtin_FIFO
+  deriving (Show, Generic)
+
+data DcConfig depth = DcConfig
+  { dcDepth          :: !depth
+  , dcImplementation :: !DcImplementation
+  , dcReadMode       :: !ReadMode
+  , dcReadDataCount  :: !Bool
+  , dcWriteDataCount :: !Bool
+  }
+  deriving (Show, Generic)
+
+defConfig :: KnownNat depth => DcConfig (SNat depth)
+defConfig = DcConfig
+  { dcDepth = SNat
+  , dcImplementation = Independent_Clocks_Block_RAM
+  , dcReadMode = Standard_FIFO
+  , dcReadDataCount = True
+  , dcWriteDataCount = True
+  }
+
+data FifoState depth n = FifoState
+  { hsQueue      :: Vec (2^depth) (Maybe (BitVector n))
+  , relativeTime :: Int
+  } deriving Show
```

Comment:
I think we had a slight miscommunication @vmchale :-). The types used in `dcFifo#` are good as they are: they will never be seen by Clash because of the BlackBox implementation. (See the ANN pragmas.) That means the internals of the FIFO can be implemented any way you like: `Seq` is a very good choice for a data structure as it has great big O complexities for appending/prepending/popping. Using `Vec` would be a step backwards here! Luckily you kept the commits separate, so it should be easy to revert.